### ISSUE TITLE
Dart2js bug: Can't extend ListMixin and call super.sort().

### ISSUE THREAD

**Issue by terrier989** [dart-lang/sdk#37494]
In SDK 2.4.0 and SDK 2.5.0-dev.1.0, this reproduces the error:
```dart
import 'dart:collection';
import 'package:test/test.dart';

void main() {
  test("Sorting a list", () {
    final list = Example();
    list._list.addAll([1,2,3]); // In SDK 2.5 the list must be non-empty.
    list.sort();
  });
}

class Example<T> extends ListBase<T> {
  final _list = <T>[];

  @override
  operator [](int index) => _list[index];

  @override
  operator []=(int index, T value) {
    _list[index] = value;
  }

  @override
  int get length => _list.length;

  @override
  set length(int value) {
    _list.length = value;
  }

  @override
  void sort([int compare(T a, T b)]) {
    super.sort(compare); // This super call causes the problem
  }
}
```
Which calls _dart:collection_ [ListMixin.sort(...)](https://github.com/dart-lang/sdk/blob/master/sdk/lib/collection/list.dart):
```dart
abstract class ListMixin<E> implements List<E> {
  // ...
  void sort([int compare(E a, E b)]) {
    Sort.sort(this, compare ?? _compareAny);
  }
}
```
Which calls _dart:internal_ [Sort.sort(...)](https://github.com/dart-lang/sdk/blob/master/sdk/lib/internal/sort.dart):
```dart
class Sort {
  // ...
  
  static void sort<E>(List<E> a, int compare(E a, E b)) {
    _doSort(a, 0, a.length - 1, compare); // LINE 33
  }

  // ...

  static void _doSort<E>(List<E> a, int left, int right, int compare(E a, E b)) {
    if ((right - left) <= _INSERTION_SORT_THRESHOLD) {
      _insertionSort(a, left, right, compare); // LINE 58 (SDK 2.4 ERROR)
    } else {
      _dualPivotQuicksort(a, left, right, compare);
    }
  }

  static void _insertionSort<E>(List<E> a, int left, int right, int compare(E a, E b)) {
    for (int i = left + 1; i <= right; i++) {
      var el = a[i]; // LINE 67 (SDK 2.5 error)
      int j = i;
      while ((j > left) && (compare(a[j - 1], el) > 0)) {
        a[j] = a[j - 1];
        j--;
      }
      a[j] = el;
    }
  }
}
```

The test passes in VM, but in Dart2js you get an error.

In SDK 2.4 the error is:
```
org-dartlang-sdk:///sdk/lib/_internal/js_runtime/lib/js_helper.dart 1274:37   Object.wrapException
org-dartlang-sdk:///sdk/lib/_internal/js_runtime/lib/js_rti.dart 471:3        Object.assertSubtype
org-dartlang-sdk:///sdk/lib/internal/sort.dart 58:22                          Object.Sort._doSort
org-dartlang-sdk:///sdk/lib/internal/sort.dart 33:5                           Example.sort
test/example.dart 33:11                                                       main.<fn>
package:test_api                                                              <fn>

TypeError: 0: type 'JSInt' is not a subtype of type 'List<dynamic>
```

In SDK 2.5, the error is a bit different, but essentially the same:
```
org-dartlang-sdk:///sdk/lib/internal/sort.dart 67:16                          Object.Sort._insertionSort
org-dartlang-sdk:///sdk/lib/internal/sort.dart 58:7                           Object.Sort._doSort
org-dartlang-sdk:///sdk/lib/internal/sort.dart 33:5                           Example.sort
test/example.dart 33:11                                                       main.<fn>
package:test_api                                                              <fn>

NoSuchMethodError: method not found: '$index' (t1.$index is not a function)
```

Conclusion:
  * The error occurs because local `a` in `_insertionSort(a, ...` should be `List`, but some bug in Dart2js makes it become one of the `int` locals.

**Comment by vsmenon** [comment]
Poked at this a bit.  Dart2JS is generating the following:

```
  Y.Example.prototype = {
    $index: function(_, index) {
      return J.$index$asx(this._list, H.intTypeCheck(index));
    },
    $indexSet: function(_, index, value) {
      J.$indexSet$ax(this._list, H.intTypeCheck(index), H.assertSubtypeOfRuntimeType(value, H.getTypeArgumentByIndex(this, 0)));
    },
    get$length: function(_) {
      return J.get$length$asx(this._list);
    },
    set$length: function(_, value) {
      J.set$length$asx(this._list, H.intTypeCheck(value));
    },
    sort$0: function(_) {
      this.super$ListMixin$sort(0, null);
    }
  };
```

In `sort$0`, the target of `this.super$ListMixin$sort` expects a `List` but gets `0` - hence the type check failure.  If I hand replace the `0` with `this._list`, it runs.

FYI - in case there is an easy fix here.
@rakudrama @fishythefish

**Comment by rakudrama** [comment]
`this.super$ListMixin$sort(0, null);` should be `this.super$ListMixin$sort(this, null);`.

**Comment by rakudrama** [comment]
Fix available: https://dart-review.googlesource.com/c/sdk/+/109553

**Comment by rakudrama** [comment]
Fixed by a089199b9337aa693f932e23dd11769f39c36d7a
